compiler/resolution/functionResolution.cpp

```diff
@@ -588,20 +594,128 @@ hasUserAssign(Type* type) {
+{
+  ArgSymbol* arg = newFn->getFormal(1);
+
+  if (AggregateType* ct = toAggregateType(arg->type)) {
+    if (isUserDefinedRecord(ct) &&
+        ct->initializerStyle == DEFINES_INITIALIZER) {
+      // If the user has defined any initializer,
+      // initCopy function should call the copy-initializer.
+      //
+      // If no copy-initializer exists, we should generate a call
+      // to a dummy init-copy function that generates an error
+      // if it remains in the AST after callDestructors
+      // (since callDestructors can remove some initCopy calls,
+      //  we'd like types that cannot be copied to survive
+      //  compilation until callDestructors has a chance to
+      //  remove those calls)).
+
+      // Go ahead and instantiate the body now so we can fix
+      // it up completely...
+      instantiateBody(newFn);
+
+      Symbol* memeTmp = newTemp(ct);
+      DefExpr* def = new DefExpr(memeTmp);
+      newFn->insertBeforeReturn(def);
+      CallExpr* initCall = new CallExpr("init", arg, memeTmp);
+      def->insertAfter(initCall);
+
+      FnSymbol* initFn = tryResolveCall(initCall);
+
+      if (initFn == NULL) {
+        // No copy-initializer could be found
+        def->remove();
+        initCall->remove();
+        newFn->addFlag(FLAG_ERRONEOUS_INITCOPY);
+      } else {
+        // Replace the other setting of the return-value-variable
+        // with what we have now...
+
+        // find the RVV
+        Symbol* retSym = newFn->getReturnSymbol();
+
+        // Remove other PRIM_MOVEs to the RVV
+        for_alist(stmt, newFn->body->body) {
+          if (CallExpr* callStmt = toCallExpr(stmt))
+            if (callStmt->isPrimitive(PRIM_MOVE)) {
+              SymExpr* se = toSymExpr(callStmt->get(1));
+              INT_ASSERT(se);
+              if (se->symbol() == retSym)
+                stmt->remove();
+            }
+        }
+
+        // Set the RVV to the copy
+        newFn->insertBeforeReturn(new CallExpr(PRIM_MOVE, retSym, memeTmp));
+
+        gdbShouldBreakHere();
+      }
+      return true;
+    }
+  }
+  return false;
+}
+
 
 static void
 resolveAutoCopyEtc(Type* type) {
 
   // resolve autoCopy
-  if (autoCopyMap.get(type) == NULL) {
+  if (!hasAutoCopyForType(type)) {
+
+    const char* copyFnToCall = "chpl__autoCopy";
+
+    if (isUserDefinedRecord(type) &&
+        !type->symbol->hasFlag(FLAG_TUPLE) &&
+        !isRecordWrappedType(type) &&
+        !isSyncType(type) &
+        !isSingleType(type) &&
+        !type->symbol->hasFlag(FLAG_EXTERN)) {
+      // Just use 'chpl__initCopy' instead of 'chpl__autoCopy'
+      // for user-defined records. This way, if the type does not
+      // support copying, the autoCopyMap will store a function
+      // marked with FLAG_ERRONEOUS_INITCOPY. Additionally, user-defined
+      // records shouldn't be defining chpl__initCopy or chpl__autoCopy
+      // and certainly shouldn't rely on the differences between the two.
```

Comment:
By removing the default auto-copy (that just calls init-copy), we enable the FLAG_ERRONEOUS_INIT_COPY strategy to work if an auto-copy is added later for a type that does not support copying. By putting something there (instead of NULL), we can let code used to working with autoCopyMap still function, but also get an appropriate error.

*at the same time* it's not appropriate for user records to have both autoCopy and initCopy. They just have copy initializers and the compiler currently uses autoCopy and initCopy as ways to call the copy initializer. I don't think user records should have either autoCopy or initCopy. autoCopy/initCopy are merely implementation concerns.